### ISSUE TITLE
Replace `for_each_mut` (internal iteration) with `iter_mut` (external iteration) in the AMT

### ISSUE THREAD

**Issue by Stebalien** [filecoin-project/ref-fvm#1996]
The iterator will have to yield a smart pointer that:

1. Can be dereferenced immutably with no affect.
2. Can be dereferenced mutably, marking the value as changed.
3. Can be deleted (e.g., by calling some form of `ptr.delete()` method). This method should _consume_ the smart pointer (possibly returning the inner value).

This will get rid of the last "internal" iteration feature.

**Comment by PhantomOz** [comment]
can i work on this?

**Comment by Stebalien** [comment]
Absolutely! Take a look at how we currently use it [here](https://github.com/filecoin-project/builtin-actors/blob/a0e34d22665ac8c84f02fea8a099216f29ffaeeb/actors/miner/src/bitfield_queue.rs#L63-L85).

**Comment by rjan90** [comment]
Hey @PhantomOz! Are you still interested in working on this, or should I unassign you?

**Comment by rjan90** [comment]
Unassigning since I have not hear back yet.

**Comment by PhantomOz** [comment]
I apologize didn't notice that i was assigned till now, Please reassign so that I can work on it @rjan90

**Comment by rjan90** [comment]
No worries! Reassigned to you now 👍

**Comment by rjan90** [comment]
Hey @PhantomOz! Do you need some additional guidance on this issue ticket? How is the progress on going on this?

**Comment by PhantomOz** [comment]
@rjan90 Yes I do

**Comment by Sahilgill24** [comment]
Hey **@rjan90** , If the issue is still open to be worked upon , Can i give it a try ?

**Comment by rjan90** [comment]
Hey @Sahilgill24! Yes, it is still open to be worked on. Go ahead and give it a try, and feel free to ask any questions you may have

**Comment by Sahilgill24** [comment]
Hey @rjan90 
My understanding of the issue is that currently  `for_each_mut()` is based on internal iteration and works like this  ```amt.for_each_mut(f)```  where the function `for_each_mut` calls the `for_each_while_mut()` inside of it  , If The function `iter_mut()` is based on  external iteration so it may look like this  
```
for i in amt.iter_mut() { 
       *(i.derefmut()).for_each_while_mut(f)
 }
``` 
 where `i` is the custom smart pointer and `f` is the closure. 
Is this the correct approach or am i doing something wrong ?

**Comment by Stebalien** [comment]
Yes, assuming the smart pointer `i` implements `DerefMut` (and `Deref`).

But note: this is a good first issue because it's well specified, not because it's easy. You're going to have to play some games with `Rc` and `RefCell` because the smart pointer cannot borrow from the iterator.

**Comment by Sahilgill24** [comment]
Hey @Stebalien  , so I was able to Implement the `iter_mut` function and successfully write the test for it but I have used `Box::leak()` which holds a value for the entire runtime but also causes leak in the memory , I will try to implement the same using `Rc` and `Refcell` also once . 

```
    // Amtptr is the custom smart pointer 
    pub fn iter_mut<'a>(&mut self) -> impl Iterator<Item = (u64, Amtptr<'a, V>)>
    where
        V: 'a + Clone,
    {
        let mut handles = Vec::new();
        let mut iterable = Vec::new();

        self.for_each_while_mut(|idx, val| {
            let val2 = val.clone();
            handles.push((idx, val2));

            Ok(true)
        });

        for i in handles {
            let tempval = Box::leak(Box::new(i.1));
            iterable.push((i.0, Amtptr::new(tempval)));
        }
        iterable.into_iter()
    }
```
This function can be used like this 
```
    for ptr in new_amt.iter_mut() {
        let current_idx = ptr.0;
        let mut val = ptr.1;
        f(current_idx, val.deref_mut()).unwrap();
        // f is a closure , which we can define earlier

    }
```

**Comment by Sahilgill24** [comment]
Hey @Stebalien , I was able to implement `iter_mut` using `Rc` and `RefCell`, but I ran into some trouble implementing `DerefMut` for them. However, instead of using the `deref_mut()` function, we can use this alternative approach, and it passes the tests as well. Compared to the `Box` approach I mentioned earlier, this does not leak memory and doesn't involve any lifetime complexities .

```
for ptr in new_amt.iter_mut() {
    let current_idx = ptr.0;
    let mut val = ptr.1;
    let mut val_ref = val.get_mut();
    let mut vale = ValueMut::new(&mut *val_ref);
    f(current_idx, &mut vale).unwrap();
 }
```
This is the smart pointer 

```
pub struct Amtptr<T> {
    value: Rc<RefCell<T>>,
    changed: bool,
}

impl<T> Amtptr<T> {
    fn new(value: T) -> Self {
        Self {
            value: Rc::new(RefCell::new(value)),
            changed: false,
        }
    }
    
    fn get_mut(&mut self) -> RefMut<'_, T> {
        self.changed = true;
        self.value.borrow_mut()
    }
}
```

**Comment by Sahilgill24** [comment]
Hey guys @Stebalien @rjan90 , could you please reply to the above comment 
Thanks !

**Comment by rjan90** [comment]
> Also shall I make a draft PR for you to have a better look at the code ?

Hey @Sahilgill24! Apologies for the delayed response. Feel free to open a draft PR so that we can take a closer look at the changes.  I would need to page in @rvagg  with regards to your technical questions here though.

**Comment by Sahilgill24** [comment]
Thanks a lot @rjan90 , made the draft PR #2178 .

**Comment by Sahilgill24** [comment]
hey @rvagg , so while trying out the issue  , I found that there was no direct way to make a iterator from the `for_each_while_mut` function because we cannot borrow a mutable reference, so my thought was to make a cloned copy of the AMT where all the changes would happen as i made in the then going to each node to do the changes in the original AMT ,but now as I thought about it we could add a function in the node to return a mutable reference to the value at a certain index and from that we may be able to create the external iterator . I can't think of any other way to do it directly . 

https://github.com/filecoin-project/ref-fvm/pull/2178 , I had given the cloning approach a try here . Could you tell which approach is better or is there anything else straightforward method that I missed .

**Comment by Sahilgill24** [comment]
hey @rvagg , any suggestion on this , Thanks !!
